src/render.asm

```diff
@@ -1,110 +1,99 @@
+// These functions are in .data because they may be called from both cores.
 .section .data
 
-.global tmds_scan
-.type tmds_scan,%function
+// r0: command list in direct threaded format
+// r1: output buffer
+// r2: y
+.global render_engine
+.type render_engine,%function
 .thumb_func
-// r0: scan list in direct threaded format
-// r1: input buffer
-// r2: output buffer
-// r3: stride (of output buffer)
-tmds_scan:
+render_engine:
     push {r4, r5, r6, r7}
-    mov r4, r8
-    mov r5, r9
-    mov r6, r10
-    push {r4, r5, r6}
+    movs r3, #0 // outw
 
-    // operation, 2x args
-    // should count be single pixels or double?
-    ldmia r0!, {r4, r5, r6}
+    ldmia r0!, {r4, r5, r6, r7}
     bx r4
 
-.global tmds_scan_stop
-.type tmds_scan_stop,%function
+.global render_stop
+.type render_stop,%function
 .thumb_func
-tmds_scan_stop:
-    subs r0, #8
-    pop {r4, r5, r6}
-    mov r8, r4
-    mov r9, r5
-    mov r10, r6
+render_stop:
     pop {r4, r5, r6, r7}
     bx lr
 
-// args: count tmds_blue tmds_green tmds_red
-// Not sure we'll keep this.
-.global tmds_scan_solid_tmds
-.type tmds_scan_solid_tmds,%function
+// args: input, stride, [rs1: u8, ls: u8, rs0: u8]
+.global render_blit_simple
+.type render_blit_simple,%function
 .thumb_func
-tmds_scan_solid_tmds:
-    mov r8, r1
-    lsls r5, #2
-    adds r4, r2, r5
-    // ip is actual end of output
-    mov ip, r4
-    lsls r5, #28
-    lsrs r5, #28
-    adds r4, r2, r5
-    // r10 is end of fractional part (may be == r2)
-    mov r10, r4
-    adds r7, r2, r3 // beginning of green row
-    cmp r2, r10
-    beq 2f
-1:
-    stmia r2!, {r6}
-    cmp r2, r10
-    bne 1b
-    cmp r2, ip
-    beq 4f
-2:
-    mov r5, r6
-3:
-    stmia r2!, {r5, r6}
-    stmia r2!, {r5, r6}
-    cmp r2, ip
-    bne 3b
-4:
+render_blit_simple:
+    muls r6, r2
+    ldr r4, [r5, r6]
+    lsrs r4, r7
+    lsrs r7, #8
+    lsls r4, r7
+    lsrs r7, #8
+    lsrs r4, r7
+    orrs r3, r4
+    ldmia r0!, {r4, r5, r6, r7}
+    bx r4
+
+// args: input, stride, [rs1: u8, ls: u8, rs0: u8]
+.global render_blit_out
```

Comment:
What operation does the `*_out` version do that is different than the normal version? Would out make sense as its own operation? or is the cost of another function call/moving registers around not worth it?

Reply:
The difference is outputting a 32-bit word (`stmia r1!, {r3}`) and clearing it to start a new word. My main concern here is that each op is 16 bytes of RAM, and for a text-heavy application (terminal etc) the RAM footprint of the display lists could be a limiting factor. It would also take a few more cycles to have a separate op, but I'm not as worried about that.